configmap/hash-gen/main.go

```diff
@@ -0,0 +1,133 @@
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+
+	"gopkg.in/yaml.v3"
+	"knative.dev/pkg/configmap"
+)
+
+func main() {
+	fileName := os.Args[1]
+	if err := processFile(fileName); err != nil {
+		log.Fatal(err)
+	}
+}
+
+// processFile reads the ConfigMap manifest from a file and adds or updates the label
+// containing the checksum of it's _example data if present.
+func processFile(fileName string) error {
+	in, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		return fmt.Errorf("failed to read file: %w", err)
+	}
+
+	out, err := process(in)
+	if out == nil || err != nil {
+		return err
+	}
+
+	if err := ioutil.WriteFile(fileName, out, 0644); err != nil {
+		return fmt.Errorf("failed to write file: %w", err)
+	}
+	return nil
+}
+
+// process processes a YAML file's bytes and adds or updates the label containing
+// the checksum of it's _example data if present.
+func process(data []byte) ([]byte, error) {
+	var doc yaml.Node
+	if err := yaml.Unmarshal(data, &doc); err != nil {
+		return nil, fmt.Errorf("failed to parse YAML: %w", err)
+	}
+	content := doc.Content[0]
+
+	example := traverse(content, "data", configmap.ExampleKey)
+	if example == nil {
+		return nil, nil
+	}
+
+	labels := traverse(content, "metadata", "labels")
+	if labels == nil {
+		// TODO(markusthoemmes): Potentially handle missing metadata and labels?
+		return nil, errors.New("'metadata.labels' not found")
+	}
+
+	checksum := fmt.Sprintf("%d", configmap.Checksum(example.Value))
+	existingLabel := value(labels, configmap.ExampleChecksumLabel)
+	if existingLabel != nil {
+		existingLabel.Value = checksum
+	} else {
+		labels.Content = append(labels.Content,
+			strNode(configmap.ExampleChecksumLabel),
+			strNode(checksum))
+	}
+
+	var buffer bytes.Buffer
+	buffer.Grow(len(data))
+	encoder := yaml.NewEncoder(&buffer)
+	encoder.SetIndent(2)
+	if err := encoder.Encode(&doc); err != nil {
+		return nil, fmt.Errorf("failed to encode YAML: %w", err)
+	}
+	return buffer.Bytes(), nil
+}
+
+// traverse traverses the YAML nodes' children using the given path keys. Returns nil if
+// one of the keys can't be found.
+func traverse(parent *yaml.Node, path ...string) *yaml.Node {
+	if parent == nil || len(path) == 0 {
+		return parent
+	}
+	tail := path[1:]
+	child := value(parent, path[0])
+	return traverse(child, tail...)
+}
+
+// value returns the value of the current node under the given key. Returns nil if not
+// found.
+func value(parent *yaml.Node, key string) *yaml.Node {
+	for i := range parent.Content {
+		if parent.Content[i].Value == key {
+			// The Content array of a yaml.Node is just an array of more nodes. The keys
+			// to each field are therefore a string node right before the relevant field's
+			// value, hence we need to return the next value of the array to retrieve
+			// the value under a key.
+			if len(parent.Content) < i+1 {
```

Comment:
```suggestion
			if len(parent.Content) < i+2 {
```
to read index `i+1` the length must be at least `i+2`